### PR TITLE
fix: missing vmnic info with `r/cluster` import

### PR DESCRIPTION
**Pull request by burnsjared0415**
In order to have a good experience with our community, we recommend that you read the [contributing guidelines](https://github.com/vmware/terraform-provider-vcf/blob/main/CONTRIBUTING.md) for making a pull request.

**Summary of Pull Request**

Fix missing `vmnic` info with `r/cluster` import.

**Type of Pull Request**

- [x] This is a bug fix.
- [ ] This is an enhancement or feature.
- [ ] This is a code style/formatting update.
- [ ] This is a documentation update.
- [ ] This is a refactoring update.
- [ ] This is a chore update
- [ ] This is something else.
      Please describe:

**Related to Existing Issues**

Ref: #210 

**Test and Documentation Coverage**

For bug fixes or features:

- [x] Tests have been completed.
- [ ] Documentation has been added/updated.

**Breaking Changes?**

- [ ] Yes, there are breaking changes.
- [x] No, there are no breaking changes.

### CODE REVIEW

File: internal/cluster/cluster_operations.go [hunk 1]
```diff
@@ -390,9 +390,7 @@ func ImportCluster(ctx context.Context, data *schema.ResourceData, apiClient *cl
 		WithTimeout(constants.DefaultVcfApiCallTimeout)
 	getClusterParams.ID = clusterId
 	clusterResult, err := apiClient.Clusters.GetCluster(getClusterParams)
-	if err != nil {
-		return nil, err
-	}
+
```
Comment (tenthirtyam) [inline internal/cluster/cluster_operations.go:393]: This should be added back. 

```suggestion
    if err != nil {
		return nil, err
	}

```

**Review by tenthirtyam — CHANGES_REQUESTED** [review]
Just a small change to add it back.

Could you also share the result and state in the testing section?

**Review by tenthirtyam — DISMISSED** [review]
LGTM

**Review by spacegospod — DISMISSED** [review]
Looks good

File: internal/cluster/host_spec_subresource.go [hunk 2]
```diff
@@ -110,6 +123,16 @@ func FlattenHost(host *models.Host) *map[string]interface{} {
...
 	}
+	if len(host.PhysicalNics) > 0 {
+		var physicalNics []map[string]interface{}
+		for _, nic := range host.PhysicalNics {
+			nicMap := make(map[string]interface{})
+			nicMap["name"] = nic.DeviceName
+			nicMap["mac_address"] = nic.MacAddress
+			physicalNics = append(physicalNics, nicMap)
+		}
+		result["vmnic"] = physicalNics
```
Comment (dimitarproynov) [inline internal/cluster/host_spec_subresource.go:134]: Can we add one line in the test to verify we're setting the "vmnic" property?

**Review by dimitarproynov — CHANGES_REQUESTED** [review]

File: internal/cluster/host_spec_subresource.go (hunk 2 above)
Comment (burnsjared0415) [reply internal/cluster/host_spec_subresource.go:134]: "availability_zone_name": "",
                "host_name": "xxxxxx",
                "id": "08bf1760-a554-4d10-a665-e571d79c1f1d",
                "ip_address": "10.156.137.1",
                "license_key": "",
                "password": "",
                "serial_number": "",
                "ssh_thumbprint": "",
                "username": "",
                "vmnic": [
                  {
                    "mac_address": "00:50:56:85:03:bc",
                    "name": "vmnic0"
                  },
                  {
                    "mac_address": "00:50:56:85:0d:33",
                    "name": "vmnic1"
                  },
                  {
                    "mac_address": "00:50:56:85:a2:d2",
                    "name": "vmnic2"
                  },
                  {
                    "mac_address": "00:50:56:85:66:59",
                    "name": "vmnic3"
                  }
                ]
              },

Comment (dimitarproynov) [reply internal/cluster/host_spec_subresource.go:134]: Hi, I ment the E2E test (here in Terraform it is called acceptance test)

Comment (burnsjared0415) [reply internal/cluster/host_spec_subresource.go:134]: here is the results
Running tool: /usr/local/bin/go test -timeout 30s -run ^TestAccDataSourceVcfCluster$ github.com/vmware/terraform-provider-vcf/internal/provider

ok  	github.com/vmware/terraform-provider-vcf/internal/provider	0.947s

File: internal/cluster/cluster_operations.go [hunk 3]
```diff
@@ -390,9 +391,7 @@ func ImportCluster(ctx context.Context, data *schema.ResourceData, apiClient *cl
 		WithTimeout(constants.DefaultVcfApiCallTimeout)
 	getClusterParams.ID = clusterId
 	clusterResult, err := apiClient.Clusters.GetCluster(getClusterParams)
-	if err != nil {
-		return nil, err
-	}
```
Comment (dimitarproynov) [inline internal/cluster/cluster_operations.go:395]: This should not be removed.

File: internal/cluster/cluster_operations.go [hunk 4]
```diff
@@ -402,6 +401,9 @@ func ImportCluster(ctx context.Context, data *schema.ResourceData, apiClient *cl
 	_ = data.Set("is_default", clusterObj.IsDefault)
 	_ = data.Set("is_stretched", clusterObj.IsStretched)
 	flattenedVdsSpecs := getFlattenedVdsSpecsForRefs(clusterObj.VdsSpecs)
+	if err != nil {
+		return nil, err
+	}
```
Comment (dimitarproynov) [inline internal/cluster/cluster_operations.go:406]: Why is this error check added here? err variable is last set at line 393.

Best practice is to return the error and exit the function immediately, not 10 instructions later.

**Review by dimitarproynov — CHANGES_REQUESTED** [review]
I don't see any tests that verify that the vmnic attribute is set on cluster's hosts.

File: internal/cluster/host_spec_subresource.go (hunk 2 above)
Comment (dimitarproynov) [reply internal/cluster/host_spec_subresource.go:134]: By "fix" i mean add code to the test that checks that the above **now** works.

The idea is: write new feature, write a new test that verifies it.
In this case there should already be a test for Cluster resource import.
Indeed, if we check the file resource_cluster_test.go we can find this: https://github.com/vmware/terraform-provider-vcf/blob/fb8d57863475d8b47c56d04c4e60ec4e33959bc5/internal/provider/resource_cluster_test.go#L557

a function that test attributes on imported cluster. We need to extend the above function to check for the newly added attributes and if they are missing, it should return an error, indicating that our newly added code doesn't work.

Comment (burnsjared0415) [reply internal/cluster/host_spec_subresource.go:134]: added it to look like this:
func TestAccDataSourceVcfCluster(t *testing.T) {
	resource.ParallelTest(t, resource.TestCase{
		PreCheck:                 func() { testAccPreCheck(t) },
		ProtoV6ProviderFactories: muxedFactories(),
		Steps: []resource.TestStep{
			{
				Config: testAccVcfClusterDataSourceConfig(
					os.Getenv(constants.VcfTestClusterDataSourceId)),
				Check: resource.ComposeTestCheckFunc(
					resource.TestCheckResourceAttrSet("data.vcf_cluster.cluster1", "name"),
					resource.TestCheckResourceAttrSet("data.vcf_cluster.cluster1", "domain_id"),
					resource.TestCheckResourceAttrSet("data.vcf_cluster.cluster1", "primary_datastore_name"),
					resource.TestCheckResourceAttrSet("data.vcf_cluster.cluster1", "primary_datastore_type"),
					resource.TestCheckResourceAttrSet("data.vcf_cluster.cluster1", "is_default"),
					resource.TestCheckResourceAttrSet("data.vcf_cluster.cluster1", "is_stretched"),
					resource.TestCheckResourceAttrSet("data.vcf_cluster.cluster1", "vds"),
					resource.TestCheckResourceAttrSet("data.vcf_cluster.cluster1", "host.0.id"),
					resource.TestCheckResourceAttrSet("data.vcf_cluster.cluster1", "host.0.host_name"),
					resource.TestCheckResourceAttrSet("data.vcf_cluster.cluster1", "host.0.ip_address"),
					resource.TestCheckResourceAttrSet("data.vcf_cluster.cluster1", "host.1.id"),
					resource.TestCheckResourceAttrSet("data.vcf_cluster.cluster1", "host.1.host_name"),
					resource.TestCheckResourceAttrSet("data.vcf_cluster.cluster1", "host.1.ip_address"),
					resource.TestCheckResourceAttrSet("data.vcf_cluster.cluster1", "host.2.id"),
					resource.TestCheckResourceAttrSet("data.vcf_cluster.cluster1", "host.2.host_name"),
					resource.TestCheckResourceAttrSet("data.vcf_cluster.cluster1", "host.2.ip_address"),
				),
			},
		},
	})
}
and getting:
Starting: /Users/burnsja/go/bin/dlv dap --listen=127.0.0.1:52527 --log-dest=3 from /Users/burnsja/Documents/GitHub/terraform-provider-vcf/internal/provider
DAP server listening at: 127.0.0.1:52527
Type 'dlv help' for list of commands.
PASS
Process 79786 has exited with status 0
Detaching

Comment (dimitarproynov) [reply internal/cluster/host_spec_subresource.go:134]: TestAccDataSourceVcfCluster test the Cluster Datasource, not the import of the Resource Cluster.

Comment (dimitarproynov) [reply internal/cluster/host_spec_subresource.go:134]: Take a look here: https://github.com/vmware/terraform-provider-vcf/blob/fb8d57863475d8b47c56d04c4e60ec4e33959bc5/internal/provider/resource_cluster_test.go#L95

Comment (burnsjared0415) [reply internal/cluster/host_spec_subresource.go:134]: ok, i see now, so there is to places that where not being tested, i worked on datasource and resouce was also missing. the code is going in now.

**Review by tenthirtyam — CHANGES_REQUESTED** [review]
@burnsjared0415 could you address the following issues from the linter and check to see if this can be addressed in the test based on Dimitar's comment.

```shell
  Error: internal/cluster/cluster_operations.go:393:17: ineffectual assignment to err (ineffassign)
  	clusterResult, err := apiClient.Clusters.GetCluster(getClusterParams)
  	               ^
  Error: internal/cluster/cluster_operations.go:406:22: ineffectual assignment to err (ineffassign)
  	flattenedHostSpecs, err := getFlattenedHostSpecsForRefs(ctx, clusterObj.Hosts, apiClient)
```

**Comment by tenthirtyam** [comment]
Rebased based on the changes to `main` recently. Needs testing again.

File: internal/cluster/host_spec_subresource.go [hunk 5]
```diff
@@ -96,6 +109,26 @@ func FlattenHost(host vcf.Host) *map[string]interface{} {
...
+	}
```
Comment (tenthirtyam) [inline internal/cluster/host_spec_subresource.go:131]: Repeated code block.

Comment (tenthirtyam) [reply internal/cluster/host_spec_subresource.go:131]: You might also consider the following:

```go
func FlattenHost(host vcf.Host) map[string]interface{} {
	result := make(map[string]interface{})
	result["id"] = host.Id
	result["host_name"] = host.Fqdn

	if host.IpAddresses != nil {
		ipAddresses := *host.IpAddresses
		if len(ipAddresses) > 0 {
			result["ip_address"] = ipAddresses[0].IpAddress
		}
	}

	if host.PhysicalNics != nil {
		var physicalNics []map[string]interface{}
		for _, nic := range host.PhysicalNics {
			nicMap := make(map[string]interface{})
			nicMap["name"] = nic.DeviceName
			nicMap["mac_address"] = nic.MacAddress
			physicalNics = append(physicalNics, nicMap)
		}
		result["vmnic"] = physicalNics
	}

	return result
}
```

**Review by tenthirtyam — CHANGES_REQUESTED** [review]
Found some repeated logic.

**Pull request closed without merging**